routing/routing_serializer.hpp

```diff
@@ -0,0 +1,202 @@
+
+struct RoutingHeader
+{
+  RoutingHeader() { Reset(); }
+
+  template <class Sink>
+  void Serialize(Sink & sink) const
+  {
+    WriteToSink(sink, m_version);
+    WriteToSink(sink, m_reserved);
+    WriteToSink(sink, m_noRestrictionCount);
+    WriteToSink(sink, m_onlyRestrictionCount);
+  }
+
+  template <class Source>
+  void Deserialize(Source & src)
+  {
+    m_version = ReadPrimitiveFromSource<uint16_t>(src);
+    m_reserved = ReadPrimitiveFromSource<uint16_t>(src);
+    m_noRestrictionCount = ReadPrimitiveFromSource<uint32_t>(src);
+    m_onlyRestrictionCount = ReadPrimitiveFromSource<uint32_t>(src);
+  }
+
+  void Reset()
+  {
+    m_version = 0;
+    m_reserved = 0;
+    m_noRestrictionCount = 0;
+    m_onlyRestrictionCount = 0;
+  }
+
+  uint16_t m_version;
+  uint16_t m_reserved;
+  uint32_t m_noRestrictionCount;
+  uint32_t m_onlyRestrictionCount;
+};
+
+static_assert(sizeof(RoutingHeader) == 12, "Wrong header size of routing section.");
+
+class RestrictionSerializer
+{
+public:
+  template <class Sink>
+  static void Serialize(RoutingHeader const & header,
+                        routing::RestrictionVec::const_iterator begin,
+                        routing::RestrictionVec::const_iterator end, Sink & sink)
+  {
+    auto const firstOnlyIt = begin + header.m_noRestrictionCount;
+    SerializeSingleType(begin, firstOnlyIt, sink);
+    SerializeSingleType(firstOnlyIt, end, sink);
+  }
+
+  template <class Source>
+  static void Deserialize(RoutingHeader const & header, routing::RestrictionVec & restrictions,
+                          Source & src)
+  {
+    DeserializeSingleType(routing::Restriction::Type::No, header.m_noRestrictionCount,
+                          restrictions, src);
+    DeserializeSingleType(routing::Restriction::Type::Only, header.m_onlyRestrictionCount,
+                          restrictions, src);
+  }
+
+private:
+  static uint32_t const kDefaultFeatureId = 0;
+
+  /// \brief Serializes a range of restrictions form |begin| to |end| to |sink|.
+  /// \param begin is an iterator to the first item to serialize.
+  /// \param end is an iterator to the element after the last element to serialize.
+  /// \note All restrictions should have the same type.
+  template <class Sink>
+  static void SerializeSingleType(routing::RestrictionVec::const_iterator begin,
+                                  routing::RestrictionVec::const_iterator end, Sink & sink)
+  {
+    if (begin == end)
+      return;
+
+    CHECK(is_sorted(begin, end), ());
+    routing::Restriction::Type const type = begin->m_type;
+
+    uint32_t prevFirstLinkFeatureId = 0;
+    for (; begin != end; ++begin)
+    {
+      CHECK_EQUAL(type, begin->m_type, ());
+
+      routing::Restriction const & restriction = *begin;
+      CHECK(restriction.IsValid(), ());
+      CHECK_LESS(1, restriction.m_featureIds.size(), ("No sense in zero or one link restrictions."));
+
+      BitWriter<FileWriter> bits(sink);
```

Comment:
BTW, why do you need to re-create bits on each iteration? Why not to create it once, before the loop?